### PR TITLE
Feature/cdr 690 new folder structure

### PR DESCRIPTION
**Pull request by stefanspiska**

### CODE REVIEW

**Review by HolgerReiseVSys — CHANGES_REQUESTED** [review]
Preliminary results

File: service/src/main/java/org/ehrbase/repository/SystemRepository.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+import org.springframework.stereotype.Repository;
+import org.springframework.transaction.annotation.Propagation;
+import org.springframework.transaction.annotation.Transactional;
+
+/**
+ * Handles DB-Access to {@link org.ehrbase.jooq.pg.tables.System}
+ * @author Stefan Spiska
+ */
+@Repository
+public class SystemRepository {
```
[the remark this answers is not in the document]
Comment (HolgerReiseVSys) [reply service/src/main/java/org/ehrbase/repository/SystemRepository.java:36]: We need to discuss if generally repositories should not be in charge of transactions; 
In this case `Propagation.MANDATORY` could make sense.

File: service/src/main/java/org/ehrbase/service/DirectoryServiceImp.java [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+
+    @Override
+    public Optional<Folder> get(UUID ehrId, @Nullable ObjectVersionId folderId, @Nullable String path) {
+
+        List<EhrFolderRecord> ehrFolderRecords;
+        if (folderId == null) {
+            ehrFolderRecords = ehrFolderRepository.getLatest(ehrId);
+        } else {
+
+            ehrFolderRecords = ehrFolderRepository.fromHistory(ehrFolderRepository.getByVersion(
```
[the remark this answers is not in the document]
Comment (HolgerReiseVSys) [reply service/src/main/java/org/ehrbase/service/DirectoryServiceImp.java:77]: `ehrFolderRepository::getByVersion` queries both tables, it just happens to use `EhrFolderHistoryRecord` as wrapper

File: service/src/main/java/org/ehrbase/service/DirectoryServiceImp.java [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+                    ehrId, Integer.parseInt(folderId.getVersionTreeId().getValue())));
```
[the remark this answers is not in the document]
Comment (HolgerReiseVSys) [reply service/src/main/java/org/ehrbase/service/DirectoryServiceImp.java:78]: This is done in `ehrFolderRepository::getByVersion`

File: UPDATING.md [hunk 4]
```diff
@@ -70,4 +70,25 @@ If you use the official Docker image you can also set this via
...
 
-see `\docker-compose.yml `
\ No newline at end of file
+see `\docker-compose.yml `
+
+## EHRbase 0.25.0
+
+### Switch to new directory structure
+
+With release 0.24.0 a new Structure to store EHR directory was introduced. There is no automatic migration of old EHR
```
Comment (HolgerReiseVSys) [inline UPDATING.md:79]: 0.25.0 ?

File: UPDATING.md [hunk 5]
```diff
@@ -70,4 +70,25 @@ If you use the official Docker image you can also set this via
...
+
+begin;
+alter table ehr.ehr drop column if exists directory;
+TRUNCATE ehr.folder, ehr.folder_hierarchy, ehr.folder_items,ehr.folder_history,ehr.folder_items_history,ehr.folder_hierarchy_history;
+alter table ehr.ehr add column directory uuid references ehr.folder(id);
+commit ;
+```
+
+If you need to migrate old EHR directory data please contact us.
+The execution of the Flyway migration script `V71__merge_duplicate_users.sql` may take its time as the duplicates are being consolidated.
```
Comment (HolgerReiseVSys) [inline UPDATING.md]: Déjà vu :D

File: base/src/main/resources/db/migration/V80__new_folder_structure.sql [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+        SELECT count(*) FROM ehr.folder INTO count_folder;
+        SELECT count(*) FROM ehr.folder_history INTO count_folder_history;
+        ALTER TABLE ehr.folder
+            ENABLE ROW LEVEL SECURITY;
+        ALTER TABLE ehr.folder_history
+            ENABLE ROW LEVEL SECURITY;
+
+        IF count_folder != 0 or count_folder_history != 0
+        THEN
+            RAISE EXCEPTION 'Can not automatically migrate directory. See UPDATING.md';
```
Comment (HolgerReiseVSys) [inline base/src/main/resources/db/migration/V80__new_folder_structure.sql]: Maybe `Systems with existing ehr directories cannot be migrated automatically; See UPDATING.md`?

File: base/src/main/resources/db/migration/V80__new_folder_structure.sql [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+
+alter table ehr.ehr
+    drop column if exists directory;
+drop table ehr.folder, ehr.folder_hierarchy, ehr.folder_items,ehr.folder_history,ehr.folder_items_history,ehr.folder_hierarchy_history,ehr.object_ref, ehr.object_ref_history;
+
+drop function ehr.admin_delete_folder(folder_id_input uuid);
+drop function ehr.admin_delete_folder_history(folder_id_input uuid);
+drop function ehr.admin_delete_folder_obj_ref_history(contribution_id_input uuid);
+
+create or replace function admin_delete_ehr_full(ehr_id_param uuid)
```
Comment (HolgerReiseVSys) [inline base/src/main/resources/db/migration/V80__new_folder_structure.sql:115]: schema prefix missing?

File: base/src/main/resources/db/migration/V80__new_folder_structure.sql [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+    returns TABLE
+            (
+                deleted boolean
+            )
+    language plpgsql
+    security definer
+    SET search_path = ehr, pg_temp
```
Comment (HolgerReiseVSys) [inline base/src/main/resources/db/migration/V80__new_folder_structure.sql:122]: is search_path needed?

File: base/src/main/resources/db/migration/V80__new_folder_structure.sql [hunk 9]
```diff
@@ -0,0 +1,288 @@
...
+    RETURN QUERY WITH
+                     -- Query IDs
+                     select_composition_ids
+                         AS (SELECT id FROM ehr.composition WHERE ehr_id = ehr_id_param),
+                     select_contribution_ids
+                         AS (SELECT id FROM ehr.contribution WHERE ehr_id = ehr_id_param),
+
+                     -- Delete data
+
+                     -- ON DELETE CASCADE:
```
Comment (HolgerReiseVSys) [inline base/src/main/resources/db/migration/V80__new_folder_structure.sql:147]: The comments need to be updated. Would be helpful to mention that, as a prerequisite, no folders must exist.

File: base/src/main/resources/db/migration/V80__new_folder_structure.sql [hunk 10]
```diff
@@ -0,0 +1,288 @@
...
+        ENABLE TRIGGER versioning_trigger;
+    ALTER TABLE ehr.status
+        ENABLE TRIGGER versioning_trigger;
+END
+$$;
+
+alter function admin_delete_ehr_full(uuid) owner to ehrbase;
+
+
+create or replace function js_ehr(uuid, text) returns json
```
Comment (HolgerReiseVSys) [inline base/src/main/resources/db/migration/V80__new_folder_structure.sql:223]: missing schema

File: service/src/main/java/org/ehrbase/service/InternalDirectoryService.java [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+import javax.annotation.Nullable;
+import org.ehrbase.api.service.DirectoryService;
+import org.ehrbase.jooq.pg.enums.ContributionChangeType;
+import org.ehrbase.jooq.pg.enums.ContributionDataType;
+import org.ehrbase.repository.ContributionRepository;
+
+/**
+ * @author Stefan Spiska
+ */
+public interface InternalDirectoryService extends DirectoryService {
```
[the remark this answers is not in the document]
Comment (stefanspiska) [reply service/src/main/java/org/ehrbase/service/InternalDirectoryService.java:33]: Setting a contribution Id should only be available in the serve and not the controller layer.

File: base/src/main/resources/db/migration/V79__new_folder_structure.sql [hunk 12]
```diff
@@ -0,0 +1,276 @@
...
+            SELECT jsonb_strip_nulls(
+                           jsonb_build_object(
+                                   '_type', 'EHR',
+                                   'ehr_id', ehr.js_canonical_hier_object_id(ehr_data.ehr_id),
+                                   'system_id', ehr.js_canonical_hier_object_id(ehr_data.system_value),
+                                   'ehr_status', ehr.js_ehr_status(ehr_data.ehr_id, server_id),
+                                   'time_created', ehr.js_dv_date_time(ehr_data.date_created, ehr_data.date_created_tz),
+                                   'contributions', contribution_json_array,
+                                   'compositions', composition_version_json_array
+                               )
```
[the remark this answers is not in the document]
Comment (stefanspiska) [reply base/src/main/resources/db/migration/V79__new_folder_structure.sql]: This is ok according to @birgerhaarbrandt

File: base/src/main/resources/db/migration/V79__new_folder_structure.sql [hunk 13]
```diff
@@ -0,0 +1,276 @@
...
+        ENABLE TRIGGER versioning_trigger;
+    ALTER TABLE ehr.status
+        ENABLE TRIGGER versioning_trigger;
+END
+$$;
+
+alter function admin_delete_ehr_full(uuid) owner to ehrbase;
+
+
+create or replace function js_ehr(uuid, text) returns json
```
[the remark this answers is not in the document]
Comment (stefanspiska) [reply base/src/main/resources/db/migration/V79__new_folder_structure.sql]: flyaway runs in the scheme ehr.

File: base/src/main/resources/db/migration/V80__new_folder_structure.sql [hunk 14]
```diff
@@ -0,0 +1,288 @@
...
+    path              TEXT[],
+    hierarchy_idx     text collate "C" not null,
+    hierarchy_idx_cap text collate "C" not null,
+    hierarchy_idx_len int              not null,
+    items             uuid[],
+    fields            jsonb,
+    namespace         TEXT default '1f332a66-0e57-11ed-861d-0242ac120002',
+    sys_version       INT              NOT NULL,
+    sys_period_lower  timestamptz      NOT NULL,
+    PRIMARY KEY (ehr_id, id),
```
Comment (HolgerReiseVSys) [inline base/src/main/resources/db/migration/V80__new_folder_structure.sql:62]: `PRIMARY KEY (ehr_id, ehr_folders_idx, row_num)` would be an alternative.
Then `id` could be optional (should then be renamed).

**Review by HolgerReiseVSys — CHANGES_REQUESTED** [review]

File: base/src/main/resources/db/migration/V80__new_folder_structure.sql (hunk 8 above)
Comment (stefanspiska) [reply base/src/main/resources/db/migration/V80__new_folder_structure.sql:122]: https://www.cybertec-postgresql.com/en/abusing-security-definer-functions/

**Review by HolgerReiseVSys — APPROVED** [review]
Some changes will be postponed to a refactoring

**Pull request merged**